# Assignment 3
**Version**: 1.0

**Student Name**: "Liwen Huang"

**Table of Content**

[Purpose](#Purpose)

[Visualization Technique (25%)](Visualization Technique (25%))

[Visualization Library (25%)](Visualization Library (25%))

[Demonstration (50%)](Demonstration (50%))

## Purpose

The purpose of this assignment is two-fold: (a) to have you engaged in independent learning with respect to a novel visual exploration method, a novel python-based information visualization toolkit, or both, and (b) to enhance our knowledge overall in the class (including mine!) by engaging in peer instruction. More specifically, I want you to write a tutorial demonstration which shows a new visualization technique (e.g. pareto charts) and/or a new visualization toolkit (e.g. bokeh, altair, seaborn, plot.ly). Feel free to do both a new technique and a new toolkit.

The purpose of using a data visualization is to tell a story, and the purpose of writing up a demonstration is to teach someone else about what you have learned so that they can use it, too. Think of this as a worked example, something like a medium blog post that you have written up and are posting so that other data scientists can learn about the potential uses of this style of data exploration and visualization. You should be concise in your wording and clear in your demonstration.


## Visualization Technique (25%)

* A narrative description of the visualization you are planning to use, describing how it works.
* A discussion of in which circumstances this visualization should and should not be used (what is is close to? What else could you consider? How does it relate to specific aspects of data?


### A narrative description of the visualization you are planning to use, describing how it works.

**Topic: Comprehensive Analysis of H1B Visa from 2016 to 2019**


The [H-1B](https://www.uscis.gov/working-in-the-united-states/temporary-workers/h-1b-specialty-occupations-dod-cooperative-research-and-development-project-workers-and-fashion) is a non-immigration visa in the United States under the Immigration and Nationality Act that allows US employers to temporarily employ foreign workers in specialty occupations. A specialty occupation requires the application of specialized knowledge and a bachelor's degree or the equivalent of work experience. However, there're an increasing number of complaints about the US companies abuse H1B visas to reduct human resource costs because citizens believe that most of H1B holders have relative low salaries comparing with Americans with same position and academic degree and those H1B holders take over citizens' jobs. Especially, when former US president, Donald Trump, suspended H1B visas denying employment permits for hundred of thousands of skilled foreign workers like programmers. Such a move hahas heightened tension between foreign skilled H1B participants and some of US citizens who have negative opinions of H1B.


1. List of top 20 employers sponsoring H-1B visa. (Bar chart)
2. List of number of application by state. (Heatmap)
3. List of top 20 job titles issuing H-1B visa. (Bar chart)
4. Average/Median annual pay of H-1B employee by year. (line chart)
5. Popular jobs by state. (Heatmap)
6. Average/Median annual pay of H-1B employee by state. (Heatmap)
7. H-1B in Data Science by year. (line chart)

Bar chart, line chart, and heatmaps will be used primarily in the analysis.

### A discussion of in which circumstances this visualization should and should not be used (what is is close to? What else could you consider? How does it relate to specific aspects of data?

* I will focus on H1B visas and employers from the US only, which means the E-3 Austrilian, H-1B1 Singapore, and H-1B1 Chile will be excluded from the dataset. 
* Since comparisons are the main goals of this topic, bar charts will be the primary tool to conduct the research. 
* Advantages of bar charts:
    - The scales and figures are easy to interpret. 
    - It summarize large complex data into an easy visual format for understanding.
* Disadvantages of bar charts:
    - Bar charts are very common and have lost impact on the readers.
* I will consider data entry error as a big issue because I found that there are a great number of typos in datasets. I believe there're still some typos or errors in the manipulated data even though I spent a lot of time on it. Besides that, some different subsidiary corporations belong to one company, which I need to convert to one company name. 

## Visualization Library (25%)
* The library you are going to use, and a background on why the library is good for this visualization. Who created it? Is it open source? How do you install it?
* A discussion of the general approach and limitations of this library. Is it declarative or procedural? Does it integrate with Jupyter? Why you decided to use this library (especially if there are other options)?


### The library you are going to use
### A discussion of the general approach and limitations of this library

* plotly Python Graphing Library

    - [x] Even though matplotlib has almost the same graphing functions as plotly graphing, I choose to use plotly because it is interactive. A mouseover of a specific point will bring up a box that has any of the information that was used to create the graph, as well as any extra information you want to include.
    
    ![image_1](assets/1.gif)
    - [x] Isolate certain information. Clicking a category in the legend of the visualization twice will isolate that category so it is the onely one we can see in the graphic.
    
    ![image_2](assets/2.gif)

    - [ ] (Limitations) Need to set the color every single time you create a new graph.

## Demonstration (50%)
* The dataset you picked and instructions for cleaning the dataset. You should pick a suitable dataset to demonstrate the technique, toolkit, and problem you are facing. 
* The quality of your demonstration. First demonstrate the basics of this approach, then show a few of the edges of how the library might be used for other cases. This is the "meat" of the assignment.


### The dataset you picked and instructions for cleaning the dataset

**Sources of datasets**

I use H-1B data from 2018 to 2020(Q1-Q4) datasets by the U.S. Department of Labor [This is the link](https://www.dol.gov/agencies/eta/foreign-labor/performance) and you can download it through Kaggle [This is the link], and H1B Dataset 2011-2017 by Kaggle [This is the link](https://www.kaggle.com/alisonggggg/h1b-dataset-20112017). 

**Instructions of manipulating datasets**

1. Remove non-H1B visa and non-US applications. Remove NaN values. 
2. Convert date column to datetime type and hourly, weekly, and monthly pay to annual pay.
3. Integrate subsidiary companies to the original corporates and correct typos.
4. Merge datasets to one dataset.

### The quality of demostration

1. Errors and typos

    Like what I talked about above, there're a lot of entry errors and typos in datasets. The limited data manipulating knowledge and skills I have used cannot eliminate these errors perfectly.
1. Need to improve the skills of using plotly

    I don't use plotly very much and am still learning. I believe there're more effective way to generate these figures than mine. For example, make subplots and merge multiple figures in one page. 
1. Need more datasets of other groups

    I need to compare these data to US citizens. For example, comparing H-1B employees and US citizens in median salaries is more persuasive that H-1B program won't hurt US employment rates.

**Import packages**

In [1]:
# pip install chart_studio

In [2]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt
%matplotlib inline
import chart_studio as py 
from plotly.offline import init_notebook_mode, iplot 
import plotly.graph_objs as go
from plotly.subplots import make_subplots
import matplotlib.pyplot as plt 
init_notebook_mode(connected=True)
from plotly import tools 
import warnings
warnings.filterwarnings("ignore")

**Read csv files**

In [3]:
def read_csv(filename):
    df_filename = pd.read_csv(filename)
    df_filename = df_filename[df_filename.VISA_CLASS == 'H-1B']
    df_filename_US = df_filename[df_filename.EMPLOYER_COUNTRY == 'UNITED STATES OF AMERICA']

    # remove nan values
    df_filename_US.isnull().sum()[df_filename_US.isnull().sum()>0]
    # remove useless columns
    columns = ['CASE_STATUS', 'EMPLOYMENT_START_DATE','EMPLOYER_NAME', 'EMPLOYER_STATE','JOB_TITLE', 'SOC_NAME','FULL_TIME_POSITION',
            'PREVAILING_WAGE','PW_UNIT_OF_PAY','WORKSITE_STATE']
    df_filename1 = df_filename_US[columns]
    df_filename2 = df_filename1[df_filename1['CASE_STATUS'].notnull()]
    df_filename2 = df_filename1[df_filename1['EMPLOYMENT_START_DATE'].notnull()]
    df_filename2 = df_filename1[df_filename1['EMPLOYER_NAME'].notnull()]
    df_filename2 = df_filename1[df_filename1['EMPLOYER_STATE'].notnull()]
    df_filename2 = df_filename1[df_filename1['JOB_TITLE'].notnull()]
    df_filename2 = df_filename1[df_filename1['SOC_NAME'].notnull()]
    df_filename2 = df_filename1[df_filename1['FULL_TIME_POSITION'].notnull()]
    df_filename2 = df_filename1[df_filename1['PREVAILING_WAGE'].notnull()]
    df_filename2 = df_filename1[df_filename1['PW_UNIT_OF_PAY'].notnull()]
    df_filename2 = df_filename1[df_filename1['WORKSITE_STATE'].notnull()]

    # convert date to pandas date time format
    df_filename2['EMPLOYMENT_START_DATE'] = pd.to_datetime(df_filename2['EMPLOYMENT_START_DATE'])

    # convert monthly, weekly and bi-weekly pay to annual pay
    for i in df_filename2.index:
        if df_filename2.loc[i, 'PW_UNIT_OF_PAY'] == 'Month':
            df_filename2.loc[i, 'PREVAILING_WAGE'] = df_filename2.loc[i, 'PREVAILING_WAGE'] * 12
        if df_filename2.loc[i, 'PW_UNIT_OF_PAY'] == 'Week':
            df_filename2.loc[i, 'PREVAILING_WAGE'] = df_filename2.loc[i, 'PREVAILING_WAGE'] * 48
        if df_filename2.loc[i, 'PW_UNIT_OF_PAY'] == 'Bi-Weekly':
            df_filename2.loc[i, 'PREVAILING_WAGE'] = df_filename2.loc[i, 'PREVAILING_WAGE'] * 24

    df_filename2.PW_UNIT_OF_PAY.replace(['Bi-Weekly','Month','Week'],['Year','Year','Year'], inplace=True)

    df_filename2.groupby(['FULL_TIME_POSITION','PW_UNIT_OF_PAY']).describe()['PREVAILING_WAGE']
    df_filename2['countvar'] = 1

    return df_filename2

In [4]:
df_2011_2017 = read_csv("assets/2011_2017.csv")
df_2018 = read_csv("assets/2018.csv")
df_2019 = read_csv("assets/2019.csv")
df_20201 = read_csv("assets/2020_1.csv")
df_20202 = read_csv("assets/2020_2.csv")
df_20203 = read_csv("assets/2020_3.csv")
df_20204 = read_csv("assets/2020_4.csv")

In [5]:
# uncommon to check each dataframe
# df_2011_2017.head()
# df_2018.head()
# df_2019.head()

**Convert hourly pay to annual pay**

* Full time: 40 hours/week
* Part time: est. 35 hours/week

In [6]:
full_2011 = df_2011_2017[(df_2011_2017.PW_UNIT_OF_PAY == 'Hour') & (df_2011_2017.FULL_TIME_POSITION == 'Y')].index
df_2011_2017.loc[full_2011,'PREVAILING_WAGE'] = df_2011_2017.loc[full_2011,'PREVAILING_WAGE'] * 1920
part_2011 = df_2011_2017[(df_2011_2017.PW_UNIT_OF_PAY == 'Hour') & (df_2011_2017.FULL_TIME_POSITION == 'N')].index
df_2011_2017.loc[part_2011,'PREVAILING_WAGE'] = df_2011_2017.loc[part_2011,'PREVAILING_WAGE'] * 1440

df_2011_2017.PW_UNIT_OF_PAY.replace('Hour','Year',inplace=True)

In [7]:
full_2018 = df_2018[(df_2018.PW_UNIT_OF_PAY == 'Hour') & (df_2018.FULL_TIME_POSITION == 'Y')].index
df_2018.loc[full_2018,'PREVAILING_WAGE'] = df_2018.loc[full_2018,'PREVAILING_WAGE'] * 1920
part_2018 = df_2018[(df_2018.PW_UNIT_OF_PAY == 'Hour') & (df_2018.FULL_TIME_POSITION == 'N')].index
df_2018.loc[part_2018,'PREVAILING_WAGE'] = df_2018.loc[part_2018,'PREVAILING_WAGE'] * 1440

df_2018.PW_UNIT_OF_PAY.replace('Hour','Year',inplace=True)

In [8]:
full_2019 = df_2019[(df_2019.PW_UNIT_OF_PAY == 'Hour') & (df_2019.FULL_TIME_POSITION == 'Y')].index
df_2019.loc[full_2019,'PREVAILING_WAGE'] = df_2019.loc[full_2019,'PREVAILING_WAGE'] * 1920
part_2019 = df_2019[(df_2019.PW_UNIT_OF_PAY == 'Hour') & (df_2019.FULL_TIME_POSITION == 'N')].index
df_2019.loc[part_2019,'PREVAILING_WAGE'] = df_2019.loc[part_2019,'PREVAILING_WAGE'] * 1440

df_2019.PW_UNIT_OF_PAY.replace('Hour','Year',inplace=True)

In [9]:
# 2020_1, full time
full_20201 = df_20201[(df_20201.PW_UNIT_OF_PAY == 'Hour') & (df_20201.FULL_TIME_POSITION == 'Y')].index
df_20201.loc[full_20201,'PREVAILING_WAGE'] = df_20201.loc[full_20201,'PREVAILING_WAGE'] * 1920

In [10]:
# 2020_1, part time
part_20201 = df_20201[(df_20201.PW_UNIT_OF_PAY == 'Hour') & (df_20201.FULL_TIME_POSITION == 'N')].index
df_20201.loc[part_20201, 'PREVAILING_WAGE'] = df_20201.loc[part_20201, 'PREVAILING_WAGE'] * 1440

df_20201.PW_UNIT_OF_PAY.replace('Hour','Year',inplace=True)

In [11]:
# 2020_2, full time
full_20202 = df_20202[(df_20202.PW_UNIT_OF_PAY == 'Hour') & (df_20202.FULL_TIME_POSITION == 'Y')].index
df_20202.loc[full_20202,'PREVAILING_WAGE'] = df_20202.loc[full_20202,'PREVAILING_WAGE'] * 1920

In [12]:
# 2020_2, part time
part_20202 = df_20202[(df_20202.PW_UNIT_OF_PAY == 'Hour') & (df_20202.FULL_TIME_POSITION == 'N')].index
df_20202.loc[part_20202, 'PREVAILING_WAGE'] = df_20202.loc[part_20202, 'PREVAILING_WAGE'] * 1440

df_20202.PW_UNIT_OF_PAY.replace('Hour','Year',inplace=True)

In [13]:
# 2020_3, full time
full_20203 = df_20203[(df_20203.PW_UNIT_OF_PAY == 'Hour') & (df_20203.FULL_TIME_POSITION == 'Y')].index
df_20203.loc[full_20203,'PREVAILING_WAGE'] = df_20203.loc[full_20203,'PREVAILING_WAGE'] * 1920

In [14]:
# 2020_3, part time
part_20203 = df_20203[(df_20203.PW_UNIT_OF_PAY == 'Hour') & (df_20203.FULL_TIME_POSITION == 'N')].index
df_20203.loc[part_20203, 'PREVAILING_WAGE'] = df_20203.loc[part_20203, 'PREVAILING_WAGE'] * 1440

df_20203.PW_UNIT_OF_PAY.replace('Hour','Year',inplace=True)

In [15]:
# 2020_4, full time
full_20204 = df_20204[(df_20204.PW_UNIT_OF_PAY == 'Hour') & (df_20204.FULL_TIME_POSITION == 'Y')].index
df_20204.loc[full_20204,'PREVAILING_WAGE'] = df_20204.loc[full_20204,'PREVAILING_WAGE'] * 1920

In [16]:
# 2020_3, part time
part_20204 = df_20204[(df_20204.PW_UNIT_OF_PAY == 'Hour') & (df_20204.FULL_TIME_POSITION == 'N')].index
df_20204.loc[part_20204, 'PREVAILING_WAGE'] = df_20204.loc[part_20204, 'PREVAILING_WAGE'] * 1440

df_20204.PW_UNIT_OF_PAY.replace('Hour','Year', inplace=True)

**Merge datasets**

In [17]:
# Merge 2020 dataset
df_2020 = pd.concat([df_20201, df_20202, df_20203, df_20204])

In [18]:
# Merge dataset 2016 - 2020
df = pd.concat([df_2011_2017, df_2018, df_2019, df_2020])

In [19]:
df['SOC_NAME'] = df['SOC_NAME'].str.upper()
df['SOC_NAME'].replace('COMPUTER SYSTEMS ANALYST','COMPUTER SYSTEMS ANALYSTS',inplace=True)
df['EMPLOYER_NAME'] = df['EMPLOYER_NAME'].str.upper()
df['EMPLOYER_NAME'].replace(['GOOGLE INC.','AIRLINE TARIFF PUBLISHING COMPANY (ATPCO)','ALBERT EINSTEIN COLLEGE OF MEDICINE OF YESHIVA UNIVERSITY','IBM','IBM CORP','IBM INDIA PRIVATE LIMITED','ALLERGAN USA, INC.','ALLERGAN SALES, LLC','AMAZON.COM SERVICES, INC.','AMAZON WEB SERVICES, INC.','AMAZON FULFILLMENT SERVICES, INC.','AMAZON FRESH LLC','AMAZON.COM.KYDC LLC','AMAZON.COM SERVICES LLC','AMDOCS INC.','AMENSYS INC','ANALOG DEVICES, INC.','ANALOG DEVICES INC.','ANTHEM INC.','COGNIZANT TECHNOLOGY SOLUTIONS U.S. CORPORATION'],['GOOGLE LLC','AIRLINE TARIFF PUBLISHING COMPANY','ALBERT EINSTEIN COLLEGE OF MEDICINE, INC.','IBM CORPORATION','IBM CORPORATION','IBM INDIA PVT LTD','ALLERGAN, INC.','ALLERGAN, INC.','AMAZON CORPORATE LLC','AMAZON CORPORATE LLC','AMAZON CORPORATE LLC','AMAZON CORPORATE LLC','AMAZON CORPORATE LLC','AMAZON CORPORATE LLC','AMDOCS, INC.','AMENSYS, INC.','ANALOG DEVICES INC','ANALOG DEVICES INC','ANTHEM, INC.','COGNIZANT TECH SOL US CORP'],inplace=True)

In [20]:
df['SOC_NAME']

0                 COMPUTER SYSTEMS ANALYSTS
1              OPERATIONS RESEARCH ANALYSTS
2                      COMPUTER PROGRAMMERS
3           COMPUTER OCCUPATIONS, ALL OTHER
4                           CREDIT ANALYSTS
                        ...                
117328    SOFTWARE DEVELOPERS, APPLICATIONS
117329        BIOCHEMISTS AND BIOPHYSICISTS
117330                  MANAGEMENT ANALYSTS
117331                 ELECTRICAL ENGINEERS
117332    SOFTWARE DEVELOPERS, APPLICATIONS
Name: SOC_NAME, Length: 2227690, dtype: object

In [21]:
df.groupby(['FULL_TIME_POSITION','PW_UNIT_OF_PAY'])['PREVAILING_WAGE'].describe()

count           mean           std  \
FULL_TIME_POSITION PW_UNIT_OF_PAY                                           
N                  Year              41552.0   62605.683368  9.203019e+05   
Y                  Year            2179277.0  106186.539354  1.834318e+06   

                                       min      25%      50%       75%  \
FULL_TIME_POSITION PW_UNIT_OF_PAY                                        
N                  Year            10440.0  37281.6  46137.6   59227.2   
Y                  Year                0.0  67059.0  81820.0  100152.0   

                                           max  
FULL_TIME_POSITION PW_UNIT_OF_PAY               
N                  Year            110882880.0  
Y                  Year            361380480.0

## Top 30 employers sponsoring H-1B visa

In [22]:
top_employee = df.groupby('EMPLOYER_NAME',as_index=False).count()
top_employee = top_employee.sort_values('countvar', ascending=False)[['EMPLOYER_NAME','countvar']][0:30]

In [23]:
topem = go.Bar(x=top_employee.EMPLOYER_NAME.values, y=top_employee.countvar.values, name='TOP 30')
layout = go.Layout(dict(title = 'TOP 30 Employers sponsoring H-1B visa', yaxis=dict(title="Num of applications")))
data = [topem]
fig = go.Figure(data, layout)
fig.update_layout(width=1500, height=850)
iplot(fig)

**Brief sum-up**

As we can see from the figure, [Cognizant Tech Solutions](https://www.cognizant.com/?cid=pse07692384610018), [Infosys Ltd](https://www.infosys.com/), [TATA Ltd](https://www.tata.com/), [Accenture LLP](https://www.accenture.com/us-en?c=acn_glb_brandexpressiongoogle_11607907&n=psgs_1020&&c=ad_usadfy17_10000001&n=psgs_Brand-%7c-US-%7c-Exact_&gclid=Cj0KCQjwrsGCBhD1ARIsALILBYqhc-84eEXMfMsTrbZzsC3fb9EV_S9fmi_bpl4DxSIF0HgAyZHMf3gaAuNKEALw_wcB&gclsrc=aw.ds), [IBM India Private LTD](https://www.ibm.com/in-en) are global outsourcing companies which take up most of H-1B quotas, 85,000 H-1B visas/year. These outsourcing companies winning tens of thousands of H-1B visas and squeezing out many American companies, including smaller start-ups.

**Top 30 employers and its H-1B status**

In [24]:
top_emp_sta = df.groupby(["EMPLOYER_NAME","CASE_STATUS"], as_index=False).count()
top_emp_sta = top_emp_sta[top_emp_sta.EMPLOYER_NAME.isin(top_employee.EMPLOYER_NAME)]

In [25]:
topem1 = go.Bar(x=top_emp_sta[top_emp_sta.CASE_STATUS == 'CERTIFIED'].sort_values('countvar', ascending=False)['EMPLOYER_NAME'].values, y=top_emp_sta[top_emp_sta.CASE_STATUS == 'CERTIFIED'].sort_values('countvar', ascending=False)['countvar'].values, name='CERTIFIED')
topem2 = go.Bar(x=top_emp_sta[top_emp_sta.CASE_STATUS == 'CERTIFIED-WITHDRAWN'].sort_values('countvar', ascending=False)['EMPLOYER_NAME'].values, y=top_emp_sta[top_emp_sta.CASE_STATUS == 'CERTIFIED-WITHDRAWN'].sort_values('countvar', ascending=False)['countvar'].values, name='CERTIFIED-WITHDRAWN')
topem3 = go.Bar(x=top_emp_sta[top_emp_sta.CASE_STATUS == 'DENIED'].sort_values('countvar', ascending=False)['EMPLOYER_NAME'].values, y=top_emp_sta[top_emp_sta.CASE_STATUS == 'DENIED'].sort_values('countvar', ascending=False)['countvar'].values, name='DENIED')
topem4 = go.Bar(x=top_emp_sta[top_emp_sta.CASE_STATUS == 'WITHDRAWN'].sort_values('countvar', ascending=False)['EMPLOYER_NAME'].values, y=top_emp_sta[top_emp_sta.CASE_STATUS == 'WITHDRAWN'].sort_values('countvar', ascending=False)['countvar'].values, name='WITHDRAWN')

data = [topem1, topem2, topem3, topem4]
layout = go.Layout(
    barmode='stack'
)
fig = go.Figure(data, layout)
fig.update_layout(width=1300, height=900)
iplot(fig)

**Brief sum-up**

The more applications, the more visas given to outsourcing companies has risen sharply.

## Number of application by state

In [26]:
dfapp = df.groupby('EMPLOYER_STATE', as_index=False).count()[['EMPLOYER_STATE','countvar']].sort_values('countvar',ascending=False)

In [27]:
numofapp = go.Bar(x=dfapp.EMPLOYER_STATE.values, y=dfapp.countvar, name='STATE')
layout = go.Layout(dict(title='NUMBER OF H-1B APPLICATION BY STATE', xaxis=dict(title='STATE'), yaxis=dict(title='Num of H-1B applications')))
data = [numofapp]
fig = go.Figure(data, layout)
fig.update_layout(width=1000, height=600)
iplot(fig)

In [28]:
fig = go.Figure(data = go.Choropleth(
    locations=dfapp.EMPLOYER_STATE,
    z=dfapp.countvar,
    locationmode='USA-states',
    colorscale='Blues',
    colorbar=dict(title='Number of H-1B applications',thickness=15)
))

layout = dict(title="2011-2020 H1B VISA APPLICATIONS (EMPLOYER STATE)",geo = dict(
            scope='usa',
            projection=dict( type='albers usa' ),
            showlakes = True,
            lakecolor = 'white'),
)
fig.update_layout(layout)
iplot(fig)

**Brief sum-up**

California, Texas, New Jersey, New York and Illinois are the top 5 major states of H-1B applications.

## Number of H-1B application by year

In [29]:
df['year'] = df.EMPLOYMENT_START_DATE.apply(lambda x:x.year)
dfyear = df.groupby('year',as_index=False).count()[['year','countvar']]

In [30]:
year = go.Scatter(
    x = dfyear.year,
    y = dfyear.countvar
)
layout = go.Layout(dict(title= "NUMBER OF APPLICATIONS PER YEAR",xaxis=dict(title="YEARS"),yaxis=dict(title="Num of applications")))
data = [year]
fig = go.Figure(data=data, layout=layout)
iplot(fig)

**Brief sum-up**

* There are 639.82k H-1B participants submitting their applications in 2019.
* A year later, only 370.937k participants submitted their applications in 2020. 

**Reasons**

* COVID-19 caused a great number of new grad students choose to return to their home countries and overseas applicants were not able to go to the US.
* COVID-19 casued increasing unemployment rates. A large amount of start-up and smaller companies closed their businesses. At the same time, the majority of corporations laid off employees, especially entry-level jobs.
* Trump government suspended the H-1B program and barred hundreds of thousands of foreigners from seeking employment in the US in June 2020. Mr. Trump blocked visas for a wide variety of jobs, including those for computer programmers and other skilled workers who enter the country under the H-1B visa, as well as those for seasonal workers in the hospitality industry, students on work-study summer programs and au pairs who arrive under other auspices.

## Top Job titles

In [31]:
dfjob = df.groupby('SOC_NAME', as_index=False).count()[['SOC_NAME','countvar']].sort_values('countvar', ascending=False)[0:10]

In [32]:
topjob = go.Bar(x=dfjob.SOC_NAME.values, y=dfjob.countvar.values, name='jobtitle')
layout = go.Layout(dict(title='TOP 10 JOBS', yaxis=dict(title='Num of H-1B applications')))
data = [topjob]
fig = go.Figure(data, layout)
iplot(fig)

**Brief sum-up**

Computer science is still the popular major for the H-1B market. 

#### Top Jobs by state

In [33]:
dum = df[["EMPLOYER_STATE","SOC_NAME"]]
dum = dum.groupby(["EMPLOYER_STATE","SOC_NAME"]).size().reset_index()
dum.columns = ['EMPLOYER_STATE', 'SOC_NAME', "COUNT"]
dum = dum.groupby(['EMPLOYER_STATE', 'SOC_NAME']).agg({'COUNT':sum})
dum = dum['COUNT'].groupby(level=0, group_keys=False)
dum = dum.apply(lambda x: x.sort_values(ascending=False).head(1))
dum = pd.DataFrame(dum).reset_index()

In [34]:
fig = go.Figure(data = go.Choropleth(
    locations=dum.EMPLOYER_STATE,
    z=dum.COUNT,
    text=dum.SOC_NAME,
    locationmode='USA-states',
    colorscale='Blues',
    colorbar = dict(title="Number of application",thickness=15)
))

layout= dict(title="Top job title in the state",geo = dict(
            scope='usa',
            projection=dict( type='albers usa' ),
            showlakes = True,
            lakecolor = 'white'),
             )
fig.update_layout(layout)

## Average annual pay by year

In [35]:
df = df[(df.FULL_TIME_POSITION == 'Y') & (df.PW_UNIT_OF_PAY == 'Year') & (df.CASE_STATUS == 'CERTIFIED-WITHDRAWN') | (df.CASE_STATUS == 'CERTIFIED') | (df.CASE_STATUS == 'Certified') | (df.CASE_STATUS == 'Certified - Withdrawn')]
df = df[(df.FULL_TIME_POSITION == 'N') & (df.PW_UNIT_OF_PAY == 'Year') & (df.CASE_STATUS == 'CERTIFIED-WITHDRAWN') | (df.CASE_STATUS == 'CERTIFIED') | (df.CASE_STATUS == 'Certified') | (df.CASE_STATUS == 'Certified - Withdrawn')]
df = df[(df.PW_UNIT_OF_PAY == 'Hour') & (df.CASE_STATUS == 'CERTIFIED-WITHDRAWN') | (df.CASE_STATUS == 'CERTIFIED') | (df.CASE_STATUS == 'Certified') | (df.CASE_STATUS == 'Certified - Withdrawn')]
df.drop(['PW_UNIT_OF_PAY'],axis=1,inplace=True)

In [36]:
df.groupby('year').agg({'PREVAILING_WAGE':'mean'})

PREVAILING_WAGE
year                   
2016.0     78121.803969
2017.0     76306.933225
2018.0     85798.691889
2019.0     89527.646521
2020.0     94581.740281
2021.0    101365.588165

In [37]:
t1 = go.Scatter(
    x=df.groupby('year').mean().index,
    y=df.groupby('year').mean().PREVAILING_WAGE
)

layout = go.Layout(dict(title= "H-1B Employees Average Annual Salaries by Year",xaxis=dict(title="YEARS"),yaxis=dict(title="AVERAGE ANNUAL SALARY")))
data = [t1]
fig = go.Figure(data=data,layout=layout)
iplot(fig)

#### Average annual pay by state

In [38]:
dum = df.groupby('EMPLOYER_STATE',as_index=False).mean()[['EMPLOYER_STATE','PREVAILING_WAGE']]

In [39]:
fig = go.Figure(data = go.Choropleth(
    locations=dum.EMPLOYER_STATE,
    z=dum.PREVAILING_WAGE,
    locationmode='USA-states',
    colorscale='agsunset',
    colorbar = dict(title='Avg salary in USD',thickness=15)
))

layout= dict(title="H-1B Employees Average Salaries by State",geo = dict(
            scope='usa',
            projection=dict( type='albers usa' ),
            showlakes = True,
            lakecolor = 'white'),
             )
fig.update_layout(layout)
iplot(fig)

## Median annual pay by year

In [40]:
df.groupby('year').agg({'PREVAILING_WAGE':'median'})

PREVAILING_WAGE
year                   
2016.0          73528.0
2017.0          70866.0
2018.0          82098.0
2019.0          85509.0
2020.0          90834.0
2021.0          97594.0

In [41]:
t1 = go.Scatter(
    x=df.groupby('year').median().index,
    y=df.groupby('year').median().PREVAILING_WAGE
)

layout = go.Layout(dict(title= " H-1B Employees Median Annual Salaries by Year",xaxis=dict(title="YEARS"),yaxis=dict(title="MEDIAN ANNUAL PAY")))
data = [t1]
fig = go.Figure(data=data,layout=layout)
iplot(fig)

**Brief sum-up**

The median annual salary💵 for H-1B employees in 2020, $\$$90,834, is much higher than the median annual salary for US citizens with at least a bachelor's degree in 2020, $\$$72,020. However, this figure can't explain this issue clearly and accurately. I need to compare salaries with market rates in the specific job. For example, a H-1B Software Programming(entry level) applicant working in Google and a US Software Programming(entry level) citizens working in Google.

In [42]:
dum = df.groupby('EMPLOYER_STATE',as_index=False).median()[['EMPLOYER_STATE','PREVAILING_WAGE']]

In [43]:
fig = go.Figure(data = go.Choropleth(
    locations=dum.EMPLOYER_STATE,
    z=dum.PREVAILING_WAGE,
    locationmode='USA-states',
    colorscale='agsunset',
    colorbar = dict(title='Median salary in USD',thickness=15)
))

layout= dict(title="H-1B Employees Median Salaries by State",geo = dict(
            scope='usa',
            projection=dict( type='albers usa' ),
            showlakes = True,
            lakecolor = 'white'),
             )
fig.update_layout(layout)
iplot(fig)

## H-1B in Data Science

In [44]:
df['DS'] = np.nan
df.DS[df['JOB_TITLE'].str.contains('DATA SCIENTIST', na=False)] = 'DATA SCIENTIST'
df.DS[df['JOB_TITLE'].str.contains('DATA ANALYST', na=False)] = 'DATA ANALYST'
df.DS[df['JOB_TITLE'].str.contains('MACHINE LEARNING', na=False)] = 'MACHINE LEARNING'
df.DS[df['JOB_TITLE'].str.contains('BUSINESS ANALYST', na=False)] = 'BUSINESS ANALYST'
df.DS[df['JOB_TITLE'].str.contains('DEEP LEARNING', na=False)] = 'DEEP LEARNING'
df.DS[df['JOB_TITLE'].str.contains('ARTIFICIAL INTELLIGENCE', na=False)] = 'ARTIFICIAL INTELLIGENCE'
df.DS[df['JOB_TITLE'].str.contains('BIG DATA', na=False)] = 'BIG DATA'
df.DS[df['JOB_TITLE'].str.contains('HADOOP', na=False)] = 'HADOOP'
df.DS[df['JOB_TITLE'].str.contains('DATA ENGINEER', na=False)] = 'DATA ENGINEER'
df['DS']= df.DS.replace(np.nan, 'Others', regex=True)

In [45]:
fig = make_subplots(rows=1, cols=2)

In [46]:
dum = df.groupby('DS',as_index=False).mean()[['DS','PREVAILING_WAGE']]

t1 =go.Bar(x=dum.DS,y=dum.PREVAILING_WAGE,name='DataScience')
layout = go.Layout(dict(title= " H-1B Employees in Data Science",xaxis=dict(title="Data Science Jobs"),yaxis=dict(title="Num of Applications")))
data = [t1]
fig = go.Figure(data=data,layout=layout)
iplot(data)

In [47]:
dum = df.groupby(['year','DS']).count().reset_index()[['year','DS','countvar']]

In [48]:
data = []
for i in dum.DS.unique():
    if i != 'Others':
        data.append(go.Scatter(x = dum[dum.DS == i].year,y= dum[dum.DS == i].countvar,name=i))

layout = go.Layout(dict(title= "H-1B Employees in Data Science",xaxis=dict(title="YEARS"),yaxis=dict(title="Number of Applications")))
        
fig = go.Figure(data,layout)    
iplot(fig) 

**Brief Summary**

uh-oh!😥

The number of application in Data Science dropped dramatically in 2020. It may be caused by COVID-19 or the Trump government. But, business analyst is the most popular job in data science all the time.